File: apps/api/app/db/models/wallet.py
import uuid
from datetime import datetime

from sqlalchemy import Boolean, Column, DateTime, ForeignKey, Integer, String
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import relationship

from app.db.base import Base, TenantMixin


class Wallet(Base, TenantMixin):
    __tablename__ = "wallets"

    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, unique=True)
    balance_cents = Column(Integer, default=0, nullable=False)
    currency = Column(String(10), default="EUR", nullable=False)
    last_topped_up = Column(DateTime, nullable=True)

    transactions = relationship("WalletTransaction", back_populates="wallet", cascade="all, delete-orphan")


class WalletTransaction(Base):
    __tablename__ = "wallet_transactions"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    wallet_id = Column(UUID(as_uuid=True), ForeignKey("wallets.id"), nullable=False)
    amount_cents = Column(Integer, nullable=False)
    type = Column(String(20), nullable=False)  # topup|debit|refund
    description = Column(String(500), nullable=True)
    reference_type = Column(String(50), nullable=True)
    reference_id = Column(UUID(as_uuid=True), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)

    wallet = relationship("Wallet", back_populates="transactions")


class ServiceCatalog(Base, TenantMixin):
    __tablename__ = "service_catalog"

    name = Column(String(255), nullable=False)
    category = Column(String(30), nullable=False)  # cantine|garderie|etude|sortie|other
    unit_price_cents = Column(Integer, nullable=False)
    billing_period = Column(String(20), nullable=False)  # daily|weekly|monthly|per_event
    active = Column(Boolean, default=True, nullable=False)

    subscriptions = relationship("ServiceSubscription", back_populates="service", cascade="all, delete-orphan")


class ServiceSubscription(Base, TenantMixin):
    __tablename__ = "service_subscriptions"

    student_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    service_id = Column(UUID(as_uuid=True), ForeignKey("service_catalog.id"), nullable=False)
    start_date = Column(DateTime, nullable=False, default=datetime.utcnow)
    end_date = Column(DateTime, nullable=True)
    days_of_week = Column(JSONB, default=list)
    status = Column(String(20), default="active")  # active|paused|cancelled

    service = relationship("ServiceCatalog", back_populates="subscriptions")

File: apps/web/src/api/wallet.ts
import client from './client';

export interface WalletData {
  id: string;
  user_id: string;
  balance_cents: number;
  currency: string;
  last_topped_up: string | null;
  recent_transactions: TransactionData[];
}

export interface TransactionData {
  id: string;
  wallet_id: string;
  amount_cents: number;
  type: string;
  description: string | null;
  reference_type: string | null;
  reference_id: string | null;
  created_at: string;
  stripe_payment_intent_id?: string | null;
}

export interface PaymentIntentData {
  client_secret: string;
  payment_intent_id: string;
}

export interface ServiceData {
  id: string;
  tenant_id: string;
  name: string;
  category: string;
  unit_price_cents: number;
  billing_period: string;
  active: boolean;
  created_at: string;
}

export interface SubscriptionData {
  id: string;
  student_id: string;
  service_id: string;
  start_date: string;
  end_date: string | null;
  days_of_week: number[];
  status: string;
  created_at: string;
}

export function getWallet() {
  return client.get<WalletData>('/v1/wallet');
}

export function topupWallet(amount_cents: number, description?: string) {
  return client.post<TransactionData>('/v1/wallet/topup', { amount_cents, description: description ?? 'Top-up' });
}

export function getTransactions(page = 1) {
  return client.get<TransactionData[]>('/v1/wallet/transactions', { params: { page } });
}

export function debitWallet(amount_cents: number, description: string) {
  return client.post<TransactionData>('/v1/wallet/debit', { amount_cents, description });
}

export function getServices() {
  return client.get<ServiceData[]>('/v1/services');
}

export function createService(data: { name: string; category: string; unit_price_cents: number; billing_period: string }) {
  return client.post<ServiceData>('/v1/services', data);
}

export function subscribeService(serviceId: string, studentId: string, daysOfWeek: number[] = []) {
  return client.post<SubscriptionData>(`/v1/services/${serviceId}/subscribe`, {
    student_id: studentId,
    days_of_week: daysOfWeek,
  });
}

export function getSubscriptions(studentId?: string) {
  return client.get<SubscriptionData[]>('/v1/wallet/subscriptions', {
    params: studentId ? { student_id: studentId } : {},
  });
}

export function cancelSubscription(subscriptionId: string) {
  return client.delete<SubscriptionData>(`/v1/wallet/subscriptions/${subscriptionId}`);
}

export function createPaymentIntent(amount_cents: number) {
  return client.post<PaymentIntentData>('/v1/wallet/create-payment-intent', { amount_cents });
}
